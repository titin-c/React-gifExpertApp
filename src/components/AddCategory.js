import React, {useState} from 'react';
import PropTypes  from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setImputValue] = useState('');

    const handleImputChange = (e) => {
        setImputValue(e.target.value);

    }

    const handleSubmit = (e) => {
        //Esto hace que no se refresque la página por defecto al enviar forms
        e.preventDefault();  


        // si la longitud , sin espacios, de valor del impurt es mayor a 0
        if(inputValue.trim().length > 0){
            // Añadimos el valor del input al array de categories
            setCategories( cats => [inputValue, ...cats]); 
            // y vaciamos el inputValue para que no se repita
            setImputValue('');
        }
        
        
    }

  return (
    <div className='buscador'>
        <div>Buscador de GIF's:</div>
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                placeholder='Inserta un Personaje...'
                value={inputValue}
                onChange={handleImputChange}
                >
            </input>
            <button type='submit' >Buscar</button>
        </form>
    </div>
  )
};

/*Es buena practica poner las funciones y valores requeridos en los componentes 
Por si otro programador coge tus componentes... 

a
Aqui indicamos que el componente AddCategory 
necesita obligatoriamente de la función  setCategory
*/

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
 