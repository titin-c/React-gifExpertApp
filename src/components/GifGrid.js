import React from 'react';
import PropTypes from 'prop-types';

import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({category}) => {

   // llamada al custom hooks useFetchGifs
    
   const {data:images, loading} = useFetchGifs( category );
  
    return (
        <>
            <h2 className='animate__animated animate__fadeInDown'>Imagenes de {category}</h2>

            { loading && <p className='cargando animate__animated animate__pulse'>Cargando... </p>} 
             <div className="card-grid">
                {
                    //mapeamos el array categories para que nos devuelva su contenido
                    //desestructuramos para que nos devuelva el id y el title
                    images.map( img =>(
                    <GifGridItem
                        key={img.id}
                    {...img}
                    />
                        //<li key={id} > {title} </li>
                    ))
                }
            </div> 
            <hr></hr>
        </>
  );
};

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}
