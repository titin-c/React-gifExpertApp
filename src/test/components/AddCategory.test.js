import React  from 'react';
import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe('Pruebas en <AddCategory />', () => {
    
    const setCategories = jest.fn();
    let wrapper = shallow( <AddCategory setCategories={setCategories}/>);

    beforeEach( ()=> {
        jest.clearAllMocks();
        wrapper = shallow( <AddCategory setCategories={setCategories}/>);

    });

    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('debe cambiar la caja de texto', () => {
      const input = wrapper.find('input');
      const value = 'Hola mundo';

      input.simulate('change', {target: { value }});
    });
    
    test('NO debe postear la información con submit vacio', () => {

        //wrappeamos el formulario y simulamos el submit
      wrapper.find('form').simulate( 'submit', { preventDefault(){} });

      expect( setCategories ).not.toHaveBeenCalled();
    });
    
    test('Debe de llamar el setCategories y limpiar la caja de texto', () => {
        
        //simular el inputChange
        const value = 'Hola pepe';
        wrapper.find('input').simulate('change', {target: { value }});

        //simular el submit
        wrapper.find('form').simulate( 'submit', { preventDefault(){} });

        //setCategories se debe de haber llamado
        expect( setCategories ).toHaveBeenCalled();
        //setCategories se debe de haber llamada 1 vez
        expect( setCategories ).toHaveBeenCalledTimes(1);
        //setCategories se debe de haber llamado con cualquier función
        expect( setCategories ).toHaveBeenCalledWith( expect.any(Function) );

        //el valor del input debe estar ''
        expect( wrapper.find('input').prop('value') ).toBe('')

      });
});
