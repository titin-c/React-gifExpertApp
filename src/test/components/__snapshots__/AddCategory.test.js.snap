// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Inserta un Personaje"
    type="text"
    value=""
  />
</form>
`;
